internal/ledgerutil/compare.go

```diff
@@ -283,51 +282,63 @@ func findAndWriteDifferences(outputDirPath string, outputFilename string, snapsh
 }
 
 // firstRecords is a struct used to hold only the earliest records up to a given limit
+// Basically a max heap with a size limit
 type firstRecords struct {
-	records       *diffRecordSlice
-	highestRecord *diffRecord
-	highestIndex  int
-	limit         int
+	records *diffRecordHeap
+	limit   int
 }
 
 func (s *firstRecords) addRecord(r *diffRecord) {
+	if s.limit == 0 {
+		return
+	}
+	// Limit not reached, can still add elements
 	if len(*s.records) < s.limit {
-		*s.records = append(*s.records, r)
-		s.setHighestRecord()
+		heap.Push(s.records, r)
+		// Limit reached, check new record against root
 	} else {
-		if r.lessThan(s.highestRecord) {
-			(*s.records)[s.highestIndex] = r
-			s.setHighestRecord()
+		// Only add records earlier than the root
+		if r.lessThan((*s.records)[0]) {
+			heap.Push(s.records, r)
+			heap.Pop(s.records)
 		}
 	}
 }
 
-func (s *firstRecords) setHighestRecord() {
-	if len(*s.records) == 1 {
-		s.highestRecord = (*s.records)[0]
-		s.highestIndex = 0
-		return
-	}
-	for i, r := range *s.records {
-		if s.highestRecord.lessThan(r) {
-			s.highestRecord = r
-			s.highestIndex = i
-		}
+// Returns list of diffRecords currently in heap sorted from earliest to latest, will empty the heap
+func (s *firstRecords) getAllRecords() []*diffRecord {
+	n := len(*s.records)
+	var t []*diffRecord
+	for i := 0; i < n; i++ {
+		r := heap.Pop(s.records)
+		t = append([]*diffRecord{r.(*diffRecord)}, t...)
 	}
+	return t
 }
 
-type diffRecordSlice []*diffRecord
+// diffRecordHeap implements heap interface where later records have higher priority than earlier records
+type diffRecordHeap []*diffRecord
 
-func (s diffRecordSlice) Len() int {
+func (s diffRecordHeap) Len() int {
 	return len(s)
 }
 
-func (s diffRecordSlice) Swap(i, j int) {
+func (s diffRecordHeap) Swap(i, j int) {
 	s[i], s[j] = s[j], s[i]
 }
 
-func (s diffRecordSlice) Less(i, j int) bool {
-	return (s[i]).lessThan(s[j])
+func (s diffRecordHeap) Less(i, j int) bool {
+	return !(s[i]).lessThan(s[j])
```

Comment:
Just a suggestion to improve the readability here -- Reverse the semantics of `lessThan` and rename it to `committedAfter` and remove the negation in front.

Reply:
I changed `lessThan` to `earlierThan` which I feel makes a lot more sense and is more readable without requiring the semantics to be changed